flink-runtime/src/main/java/org/apache/flink/runtime/state/CheckpointStreamWithResultProvider.java

```diff
@@ -0,0 +1,175 @@
+
+			StreamStateHandle streamStateHandle = outputStream.closeAndGetHandle();
+			KeyGroupsStateHandle primaryStateHandle = new KeyGroupsStateHandle(kgOffs, streamStateHandle);
+			return new SnapshotResult<>(primaryStateHandle, null);
+		}
+
+		@Override
+		public CheckpointStreamFactory.CheckpointStateOutputStream getCheckpointOutputStream() {
+			return outputStream;
+		}
+	}
+
+	/**
+	 * Implementation of {@link CheckpointStreamWithResultProvider} that creates both, the
+	 * primary/remote/jm-owned state and the secondary/local/tm-owned state.
+	 */
+	class PrimaryAndSecondaryStream implements CheckpointStreamWithResultProvider {
+
+		private static final Logger LOG = LoggerFactory.getLogger(PrimaryAndSecondaryStream.class);
+
+		private final DuplicatingCheckpointOutputStream outputStream;
+
+		public PrimaryAndSecondaryStream(DuplicatingCheckpointOutputStream outputStream) {
+			this.outputStream = Preconditions.checkNotNull(outputStream);
+		}
+
+		@Override
+		public SnapshotResult<KeyedStateHandle> closeAndFinalizeCheckpointStreamResult(
+			KeyGroupRangeOffsets offsets) throws IOException {
+
+			final StreamStateHandle primaryStreamStateHandle;
+
+			try {
+				primaryStreamStateHandle = outputStream.closeAndGetPrimaryHandle();
+			} catch (IOException primaryEx) {
+				try {
+					outputStream.close();
+				} catch (IOException closeEx) {
+					primaryEx = ExceptionUtils.firstOrSuppressed(closeEx, primaryEx);
+				}
+				throw primaryEx;
+			}
+
+			StreamStateHandle secondaryStreamStateHandle = null;
+
+			try {
+				secondaryStreamStateHandle = outputStream.closeAndGetSecondaryHandle();
+			} catch (IOException secondaryEx) {
+				LOG.warn("Exception from secondary/local checkpoint stream.", secondaryEx);
+			}
+
+			if (primaryStreamStateHandle != null) {
+
+				final KeyGroupsStateHandle primaryKeyGroupsStateHandle =
+					new KeyGroupsStateHandle(offsets, primaryStreamStateHandle);
+
+				final KeyGroupsStateHandle secondaryKeyGroupsStateHandle = secondaryStreamStateHandle != null ?
+					new KeyGroupsStateHandle(offsets, secondaryStreamStateHandle) :
+					null;
+
+				return new SnapshotResult<>(primaryKeyGroupsStateHandle, secondaryKeyGroupsStateHandle);
+			}
+
+			return null;
+		}
+
+		@Override
+		public DuplicatingCheckpointOutputStream getCheckpointOutputStream() {
+			return outputStream;
+		}
+	}
+
+	class Factory {
+
+		private static final Logger LOG = LoggerFactory.getLogger(PrimaryAndSecondaryStream.class);
+
+		public CheckpointStreamWithResultProvider create(
+			long checkpointId,
+			long timestamp,
+			CheckpointStreamFactory primaryStreamFactory,
+			LocalRecoveryDirectoryProvider secondaryStreamDirProvider) throws Exception {
+
+			CheckpointStreamFactory.CheckpointStateOutputStream primaryOut =
+				primaryStreamFactory.createCheckpointStateOutputStream(checkpointId, timestamp);
+
+			CheckpointStreamFactory.CheckpointStateOutputStream secondaryOut;
+
+			if (secondaryStreamDirProvider != null) {
```

Comment:
Why do we allow passing in `null` values? Wouldn't it be easier to have two methods for the simple and duplicated stream?

Reply:
👍